libvast/src/evaluate.cpp

```diff
@@ -0,0 +1,465 @@
+      return neg != result;
+    }
+    case meta_extractor::kind::import_time:
+      return evaluate(data{slice.import_time()}, op, rhs);
+  }
+  __builtin_unreachable();
+}
+
+template <relational_operator Op>
+struct cell_evaluator;
+
+template <>
+struct cell_evaluator<relational_operator::equal> {
+  static bool evaluate(auto, auto) noexcept {
+    return false;
+  }
+
+  template <class LhsView, class RhsView>
+    requires requires(const LhsView& lhs, const RhsView& rhs) {
+               { lhs == rhs } -> std::same_as<bool>;
+             }
+  static bool evaluate(LhsView lhs, RhsView rhs) noexcept {
+    return lhs == rhs;
+  }
+
+  static bool evaluate(std::string_view lhs, view<pattern> rhs) noexcept {
+    return evaluate(rhs, lhs);
+  }
+
+  static bool evaluate(view<pattern> lhs, std::string_view rhs) noexcept {
+    return lhs.match(rhs);
+  }
+};
+
+template <>
+struct cell_evaluator<relational_operator::not_equal> {
+  static bool evaluate(auto lhs, auto rhs) noexcept {
+    return !cell_evaluator<relational_operator::equal>::evaluate(lhs, rhs);
+  }
+};
+
+template <>
+struct cell_evaluator<relational_operator::less> {
+  static bool evaluate(auto, auto) noexcept {
+    return false;
+  }
+
+  template <class LhsView, class RhsView>
+    requires requires(const LhsView& lhs, const RhsView& rhs) {
+               { lhs < rhs } -> std::same_as<bool>;
+             }
+  static bool evaluate(LhsView lhs, RhsView rhs) noexcept {
+    return lhs < rhs;
+  }
+};
+
+template <>
+struct cell_evaluator<relational_operator::less_equal> {
+  static bool evaluate(auto, auto) noexcept {
+    return false;
+  }
+
+  template <class LhsView, class RhsView>
+    requires requires(const LhsView& lhs, const RhsView& rhs) {
+               { lhs <= rhs } -> std::same_as<bool>;
+             }
+  static bool evaluate(LhsView lhs, RhsView rhs) noexcept {
+    return lhs <= rhs;
+  }
+};
+
+template <>
+struct cell_evaluator<relational_operator::greater> {
+  static bool evaluate(auto, auto) noexcept {
+    return false;
+  }
+
+  template <class LhsView, class RhsView>
+    requires requires(const LhsView& lhs, const RhsView& rhs) {
+               { lhs > rhs } -> std::same_as<bool>;
+             }
+  static bool evaluate(LhsView lhs, RhsView rhs) noexcept {
+    return lhs > rhs;
+  }
+};
+
+template <>
+struct cell_evaluator<relational_operator::greater_equal> {
+  static bool evaluate(auto, auto) noexcept {
+    return false;
+  }
+
+  template <class LhsView, class RhsView>
+    requires requires(const LhsView& lhs, const RhsView& rhs) {
+               { lhs >= rhs } -> std::same_as<bool>;
+             }
+  static bool evaluate(LhsView lhs, RhsView rhs) noexcept {
+    return lhs >= rhs;
+  }
+};
+
+template <>
+struct cell_evaluator<relational_operator::match> {
+  static bool evaluate(auto, auto) noexcept {
+    return false;
+  }
+
+  static bool evaluate(view<std::string> lhs, view<pattern> rhs) noexcept {
+    return rhs.match(lhs);
```

Comment:
You also dispatch to `cell_evaluator<equals>` to have only one place in the code where you dispatch to the concrete implementation.